camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/policy/GroupsChangePolicyOsgiTest.java

```diff
@@ -0,0 +1,472 @@
+    }
+
+    /** Test effector that adds a sensor with a test value. */
+    public static final class EffectorForThisTest extends EntityInitializers.InitializerPatternWithConfigKeys implements Serializable {
+
+        public static final String NAME = "testEffector";
+        public static final String SENSOR = "testSensor";
+        public static final String SENSOR_VALUE = "testSensorValue";
+
+        @Override
+        public void apply(final EntityLocal entity) {
+            ((EntityInternal) entity).getMutableEntityType().addEffector(Effectors.effector(Void.class, NAME).impl(new EffectorBody<Void>() {
+                @Override
+                public Void call(ConfigBag parameters) {
+                    entity().sensors().set(Sensors.newStringSensor(SENSOR), SENSOR_VALUE);
+                    return null;
+                }
+            }).build());
+        }
+    }
+
+    /**
+     * Tests {@link GroupsChangePolicy#INITIALIZERS} of the member while joining the {@link DynamicGroup},
+     * {@link StaticSensor} in particular.
+     */
+    @Test
+    public void testAddInitializers() throws Exception {
+
+        // Member properties, these are expected to be resolved with DSL in the member context.
+        final String MEMBER_SENSOR_VALUE = "member-sensor-value";
+
+        // Group properties, these should not preempt member properties.
+        final String GROUP_SENSOR_VALUE = "group-sensor-value";
+
+        // Config names.
+        final String SENSOR_TO_COPY_FROM = "sensor-to-copy-from";
+        final String MEMBER_SENSOR_TO_CREATE = "sensor-to-create";
+
+        // Blueprint to test.
+        String yaml = Joiner.on("\n").join(
+                "services:",
+                "- type: " + DynamicGroup.class.getName(),
+                "  brooklyn.policies:",
+                "  - type: " + GroupsChangePolicy.class.getName(),
+                "    brooklyn.config:",
+                "      group: $brooklyn:self()",
+                "      member.initializers:",
+                "      - type: " + StaticSensor.class.getName(),
+                "        brooklyn.config:",
+                "          name: " + MEMBER_SENSOR_TO_CREATE,
+                "          target.type: string",
+                "          static.value: $brooklyn:attributeWhenReady(\"" + SENSOR_TO_COPY_FROM + "\")",
+                "- type: " + TestEntity.class.getName());
+
+        final Entity app = createStartWaitAndLogApplication(yaml);
+        Asserts.assertNotNull(app);
+
+        final DynamicGroup dynamicGroup = (DynamicGroup) Iterables.getFirst(app.getChildren(), null);
+        Asserts.assertNotNull(dynamicGroup);
+
+        final TestEntity member = (TestEntity) Iterables.getLast(app.getChildren());
+        Asserts.assertNotNull(member);
+
+        // Verify that member has joined.
+        dynamicGroup.setEntityFilter(EntityPredicates.idEqualTo(member.getId()));
+        Asserts.assertEquals(dynamicGroup.getMembers().size(), 1);
+
```

Comment:
personal preference, but i would add a convenience method

```
Pair<Entity,Etity> createAppAndGetGroupAndMember(String groupChangePolicyConfig, String optionalGroupConfig) {
  String yaml = Joiner.on("\n").join(
                "services:",
                "- type: " + DynamicGroup.class.getName(),
                (Strings.isNonBlank(optionalGroupConfig) ? "  brooklyn.config:\n"+Strings.indentEachLine(optionalGroupConfig, "    ") : ""),
                "  brooklyn.policies:",
                "  - type: " + GroupsChangePolicy.class.getName(),
                "    brooklyn.config:",
                "      group: $brooklyn:self()",
                Strings.indentEachLine(groupChangePolicyConfig, "      "),
                "- type: " + TestEntity.class.getName());

        final Entity app = createStartWaitAndLogApplication(yaml);
        Asserts.assertNotNull(app);

        final DynamicGroup dynamicGroup = (DynamicGroup) Iterables.getFirst(app.getChildren(), null);
        Asserts.assertNotNull(dynamicGroup);

        final TestEntity member = (TestEntity) Iterables.getLast(app.getChildren());
        Asserts.assertNotNull(member);

        // Verify that member has joined.
        dynamicGroup.setEntityFilter(EntityPredicates.idEqualTo(member.getId()));
        Asserts.assertEquals(dynamicGroup.getMembers().size(), 1);

        return Pair.of(dynamicGroup, member);
}
```

then each test could be rewritten more concisely

```
  final String MEMBER_SENSOR_TO_CREATE = ...;
  ...

  createAppAndGetGroupAndMember(Strings.lines(
                "member.initializers:",
                "- type: " + StaticSensor.class.getName(),
                "  brooklyn.config:",
                "    name: " + MEMBER_SENSOR_TO_CREATE,
                "    target.type: string",
                "    static.value: $brooklyn:attributeWhenReady(\"" + SENSOR_TO_COPY_FROM + "\")",
  ), null);

       // Emmit dynamic attributes for both: group and its member.
        executor.submit(() -> dynamicGroup.sensors().set(Sensors.newStringSensor(SENSOR_TO_COPY_FROM), GROUP_SENSOR_VALUE));
        executor.submit(() -> member.sensors().set(Sensors.newStringSensor(SENSOR_TO_COPY_FROM), MEMBER_SENSOR_VALUE));

        // Verify that location is added to the group member with expected member properties.
        Asserts.eventually(() -> member.getAttribute(Sensors.newStringSensor(MEMBER_SENSOR_TO_CREATE)), MEMBER_SENSOR_VALUE::equals);
```

Reply:
This is as far as I see it optimised 6c19c45be964d66a89f8e6a8b339d2ba69674955